torch/csrc/jit/codegen/cuda/kernel_cache.cpp

```diff
@@ -451,77 +456,212 @@ void FusionKernelRuntime::prepareRuntimeOrder() {
   }
 }
 
-std::vector<at::Tensor> FusionKernelRuntime::runWithInput(
-    const at::ArrayRef<IValue>& inputs,
-    size_t input_id) {
-  FUSER_PERF_SCOPE("FusionKernelRuntime::runMultiKernelWithInput");
+// passing args by value, since we will be modify this
+void FusionKernelRuntime::compileAsync(KernelArgumentHolder& args_old) {
+  // only single compilation is supported at this moment.
+  std::unique_lock<std::mutex> unique_lock(mutex_, std::try_to_lock);
+  TORCH_CHECK(
+      unique_lock.owns_lock(),
+      "Calling compileAsync on a FusionKernelRuntime that's already starting a compilation thread is not supported");
+  std::unique_lock<std::mutex> unique_lock2(compiling_, std::try_to_lock);
+  TORCH_CHECK(
+      unique_lock2.owns_lock(),
+      "Calling compileAsync on a FusionKernelRuntime that's already starting a compilation thread is not supported 2");
+
+  // for some reason I can't seem to move unique_lock and it keeps using copy.
+  // auto func = [args = std::move(args_old), lock = std::move(unique_lock),
+  // this] () mutable {
+  auto func = [args = std::move(args_old), this]() mutable {
+    std::lock_guard<std::mutex> guard(compiling_);
```

Comment:
nit: is it double `lock`/`unlock`ing `compiling_` ?

Reply:
Oh I see. the new thread would wait for `compileAsync` to return... Is that correct/intended?

Reply:
That's not intended and unnecessary.

I just haven't worked out how to move the lock through to the lambda, which is really bizarre....